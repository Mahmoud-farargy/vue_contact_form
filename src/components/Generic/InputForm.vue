<template>
    <div id="inputForm" class="form-group">
        <!-- If Input-->
        <input class="input__style"
                v-if="lowerInputType === 'text'"
                :type="specificType"
                @input="changeInput"
                :defaultValue="defaultVal"
                v-model="element.val"
                :placeholder="element.placeholder ? `${element.placeholder}${element.isRequired ? ' *' : ''}`: ''"
                :id="element.label"
                :disabled="isDisabled"
                :autoFocus="isFocused"
                :name="name"
                :required="element.isRequired"
        />
        
        <!--Else if Textarea-->
        <textarea
                class="txt__area__style"
                v-else-if="lowerInputType === 'textarea'"
                @input="changeInput"
                :defaultValue="defaultVal"
                v-model="element.val"
                :placeholder="element.placeholder ? `${element.placeholder}${element.isRequired ? ' *' : ''}`: ''"
                :id="element.label"
                :disabled="isDisabled"
                :autoFocus="isFocused"
                :name="name"
                :required="element.isRequired"
        />
        <!--Otherwise-->
        <input class="input__style"
                v-else
                :type="specificType"
                @input="changeInput"
                :defaultValue="defaultVal"
                v-model="element.val"
                :placeholder="element.placeholder ? `${element.placeholder}${element.isRequired ? ' *' : ''}`: ''"
                :id="element.label"
                :disabled="isDisabled"
                :autoFocus="isFocused"
                :name="name"
                :required="element.isRequired"
        />
        <p class="text-danger" v-if="isSubmitted && ((!element.isValid && element.isRequired) || (element.val && !element.isValid && !element.isRequired))">{{isSubmitted && !element.val ? "This field is required" :element.errorMsg}}</p>
    </div>
</template>

<script>
    export default {
        props:{
            element: {
                type:Object,
                required: true
            },
            changeInput:  {
                type:Function,
                required: true
            },
            name: {
                type: String,
                required: true
            },
            inputType: {
                type: String,
            },
            specificType: {
                type: String,
            },
            defaultVal:  {
                type:String
            },
            isDisabled: {
                type: Boolean,
                default: false
            },
            isFocused: {
                type: Boolean,
                default: false
            },
            isSubmitted: {
                type: Boolean,
                default: false
            }
        },
        computed: {
            lowerInputType(){
                return this.inputType?.toLowerCase();
            }
        }
    }
</script>

<style lang="scss" scoped>
    #inputForm{
        .txt__area__style{
            resize: vertical;
            min-height:140px;
            max-height: 170px;
        }
        .txt__area__style, input.input__style{
            background: var(--shadow-white);
            width:100%;
            padding: 9px 5px 9px 14px;
            border-radius: 3px;
            display: block;
            border:none;
            margin-bottom: 18px;
            border: 1px solid;
            border-color: transparent;
            transition: all 200ms ease;
            &:focus{
                outline: none;
                border-color: var(--primary-clr) !important;
            }
            &:hover{
                border-color: var(--light-gray);
            }
        }
    }
</style>