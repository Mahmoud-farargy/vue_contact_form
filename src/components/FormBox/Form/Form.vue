<template> 
    <form id="locationMap" @submit="onFormSubmission">
        <span class="vertical__divider"></span>
        
        <div class="form--inner">
            <InputForm :element="formState.name" defaultVal="" name="name" inputType="text" specificType="text" :changeInput="changeInput" :isSubmitted="isSubmitted"/>
            <InputForm :element="formState.email" defaultVal="" name="email" inputType="text" specificType="email" :changeInput="changeInput" :isSubmitted="isSubmitted"/>
            <InputForm :element="formState.subject" defaultVal="" name="subject" inputType="text" specificType="text" :changeInput="changeInput" :isSubmitted="isSubmitted"/>
            <InputForm :element="formState.message" defaultVal="" name="message" inputType="textarea" :changeInput="changeInput"  :isSubmitted="isSubmitted"/>
            <pulse-loader class="loader" :loading="isLoading.submittingForm" color="var(--primary-clr)" size="8px"></pulse-loader>
            <input class="submit__button" :disabled="(!areRequiredInputsFilled || isLoading.submittingForm)" :class="(!areRequiredInputsFilled || isLoading.submittingForm) &&'disabled'" type="submit" role="button" value="SEND â†’"/>
        </div>
    </form>
</template>

<script>
    import InputForm from "../../Generic/InputForm.vue";
    import { notify, generateNewId } from "../../../utilities/utilities.js";
    import { mapActions, mapGetters } from "vuex";
    import axios from 'axios';
    import PulseLoader from 'vue-spinner/src/PulseLoader.vue';
    // WE COULD ALSO USE VULIDATE FOR FORM VALIDATION BUT I DECIDED TO USE MY CUSTOM VALIDATION
    export default {
        data() {
            return {
                isSubmitted: false,
                formState: {
                    name: {val: "", regex: /^.{5,50}$/, isValid: false, errorMsg: "Name should be between 5 and 50 characters.", isRequired: true, label:"name", placeholder: "Your Name"},
                    email: {val: "", regex: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/, isValid: false, errorMsg: "Please type a valid email.", isRequired: true, label:"email_address", placeholder:"Email Address"},
                    subject: {val: "", regex:  /^.{0,100}$/, isValid: false, errorMsg: "Subject should not exceed 100 characters.", isRequired: false, label:"subject", placeholder:"Subject"},
                    message: {val: "", regex: /^.{1,500}$/, isValid: false, errorMsg: "Message should not exceed 500 characters. Please type fewer words.", isRequired: true, label:"message", placeholder:"Message"},
                }
            }
        },
        components:{
            InputForm,
            PulseLoader
        },
        computed: {
            ...mapGetters("loadingState", {isLoading: "getLoading"}),
            areRequiredInputsFilled(){
               return Object.values(this.formState).every(input => input.isRequired ? input.val: true);
            }
        },
        methods:{
                ...mapActions("loadingState", ["toggleLoading"]),
                changeInput(e) {
                    const inputTarget = e.target;
                    const inputType = inputTarget.type;
                    const inputName = inputTarget.name;
                    const inputVal = inputType === "checkbox" || inputType === "radio" ? inputTarget.checked :inputTarget.value;
                    const elementState = this.formState[inputName];
                    const isValid = elementState?.regex instanceof RegExp ? (elementState?.regex?.test(inputVal)) : false;

                    this.formState = {
                        ...this.formState,
                        [inputName]: {
                            ...this.formState[inputName],
                            val: inputVal,
                            isValid
                        }
                    }
                },
                resetForm() {
                    let newState = {...this.formState};
                    this.isSubmitted = false;
                    Object.keys(this.formState).forEach(key => {
                        newState = {
                            ...newState,
                            [key]: {
                                ...newState[key],
                                val: "",
                                isValid: false
                            }
                        }
                    });
                    this.formState = newState;
                },
                onFormSubmission(x){
                    x.preventDefault();
                    this.isSubmitted = true;
                    this.toggleLoading({type: "submittingForm", val: true});
                    new Promise((resolve, reject) => {
                       
                        if(Object.values(this.formState).every(item => item.isRequired ? (item.isValid && item.val !== "") : (item.val !== "" ? item.isValid : true))){
                             axios.post("https://61603037faa03600179fb926.mockapi.io/form", {
                                name: this.formState.name.val,
                                email: this.formState.email.val,
                                subject: this.formState.subject.val,
                                message: this.formState.message.val,
                                id: generateNewId(),
                                date: new Date()
                            }).then(() => {
                                this.toggleLoading({type: "submittingForm", val: false});
                                notify({type: "success", msg: "Your message has been sent successfully"});
                                this.resetForm();
                                resolve();
                            }).catch(error => {
                                this.toggleLoading({type: "submittingForm", val: false});
                                notify({type: "error", msg: error.message || "An error has occured"});
                                reject();
                            });
                        }else{
                            reject();
                            this.toggleLoading({type: "submittingForm", val: false});
                        }
                    })
                }
            }
        }
</script>

<style lang="scss" scoped>
    #locationMap{
        position: relative;
        min-height: 150px;
        .vertical__divider{
            display: none;         
        }
        .form--inner{
            width:320px;
            .loader{
                width: 100%;
                height: 20px;
                text-align:center;
                display: block;
            }
        }
        .submit__button{
            background-color: var(--primary-clr);
            color: var(--ultra-white);
            border:none;
            padding: 9px 10px;
            text-align:center;
            width:100%;
            text-transform: uppercase;
            cursor: pointer;
            border-radius: 2px;
            margin: 15px 0;
            transition: var(--mild-transition);
            &:hover{
                filter: brightness(1.6);
            }
        }
        // responsive 67px and above
        @media only screen and (min-width: 670px){
                .vertical__divider{
                    position: absolute;
                    display: inline;
                    right: -10px;
                    top: 50%;
                    transform: translate(-50%,-50%);
                    background-color: var(--shadow-white);
                    width: 2px;
                    height: 90%;
                    border-radius: 50%;
                    margin-right: 10px;
                    padding:0;
                }
        }
          // responsive 670px and lower
        @media only screen and (max-width: 670px){
             .form--inner{
                    width:98%;
             }
        }
    }
</style>