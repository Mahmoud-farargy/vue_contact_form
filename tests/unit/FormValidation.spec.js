import { shallowMount, createLocalVue } from '@vue/test-utils';
import Form from "../../src/components/FormBox/Form/Form.vue";
import InputForm from "../../src/components/Generic/InputForm.vue";

import Vuex from 'vuex';
const localVue = createLocalVue();
localVue.use(Vuex);

describe("Form.vue && InputForm.vue", () => {
    // InputForm.vue
    const formState = {
        name: { val: "mahmoud farargy", regex: /^.{5,50}$/, isValid: false, errorMsg: "Name should be between 5 and 50 characters.", isRequired: true, label: "name", placeholder: "Your Name" },
        email: { val: "testing134@gmail.com", regex: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/, isValid: false, errorMsg: "Please type a valid email.", isRequired: true, label: "email_address", placeholder: "Email Address" },
        subject: { val: "1290454", regex: /^.{0,100}$/, isValid: false, errorMsg: "Subject should not exceed 100 characters.", isRequired: false, label: "subject", placeholder: "Subject" },
        message: { val: "test messaging", regex: /^.{1,500}$/, isValid: false, errorMsg: "Message should not exceed 500 characters. Please type fewer words.", isRequired: true, label: "message", placeholder: "Message" },
    }
    const wrapper = shallowMount(InputForm, {
        propsData: {
            element: formState.name,
            defaultVal: "",
            name: "name",
            inputType: "text",
            specificType: "text",
            changeInput: () => { },
            isSubmitted: true
        }
    });
    it("Matches a snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
    it("Validate inputs", () => {
        Object.keys(formState).forEach((key) => {
            const item = wrapper.find(`input[type="text"]`);
            const value = formState[key].val;
            item.setValue(value);
            expect(item.element.value).toBe(value);
        });
    });

    // Form.vue
    const parent = shallowMount(Form, { store, localVue });
    const submitBtn = parent.find(".submit__button");
    it("Matches a snapshot", () => {
        expect(parent).toMatchSnapshot();
    });
    it("Makes sure the submit button is disabled when inputs are empty", () => {

        expect(submitBtn.classes()).toContain("disabled");
    });

    let store = new Vuex.Store({
        state: {
            loading: {
                submittingForm: false
            }
        }
    });
    
    it("Makes sure the submit button is clickable", () => {
        submitBtn.trigger('click');
        expect(submitBtn.element.value).toContain("SEND");
    });
});