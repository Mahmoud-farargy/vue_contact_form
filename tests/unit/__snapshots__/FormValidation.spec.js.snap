// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.vue && InputForm.vue Matches a snapshot 1`] = `
<div id="inputForm" class="form-group"><input defaultvalue="" placeholder="Your Name *" id="name" name="name" required="required" type="text" class="input__style">
  <p class="text-danger">Name should be between 5 and 50 characters.</p>
</div>
`;

exports[`Form.vue && InputForm.vue Matches a snapshot 2`] = `
<form id="locationMap"><span class="vertical__divider"></span>
  <div class="form--inner">
    <inputform-stub element="[object Object]" changeinput="[Function]" name="name" inputtype="text" specifictype="text" defaultval=""></inputform-stub>
    <inputform-stub element="[object Object]" changeinput="[Function]" name="email" inputtype="text" specifictype="email" defaultval=""></inputform-stub>
    <inputform-stub element="[object Object]" changeinput="[Function]" name="subject" inputtype="text" specifictype="text" defaultval=""></inputform-stub>
    <inputform-stub element="[object Object]" changeinput="[Function]" name="message" inputtype="textarea" defaultval=""></inputform-stub> <input disabled="disabled" type="submit" role="button" value="SEND â†’" class="submit__button disabled">
  </div>
</form>
`;
